### PR TITLE
Make sure we have no args constructor to prevent usage of unsafe

### LINKED ISSUE #199: Create "no-args" constructor for GSON POJOs to prevent using "sun.misc.Unsafe"
Please read this post: https://stackoverflow.com/a/18645370/950427.

If you are not using a "no-args" constructor in Gson, it will use: `sun.misc.Unsafe`. See https://github.com/google/gson/blob/master/UserGuide.md#writing-an-instance-creator

Also, users will no longer have to keep "unsafe" for proguard: https://github.com/google/gson/blob/master/examples/android-proguard-example/proguard.cfg#L10

Same idea from https://github.com/AzureAD/azure-activedirectory-library-for-android/issues/839

### PR DESCRIPTION
**Pull request by jaredsburrows**
@addev-ashish
Closes #199

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/MsalServiceException.java [hunk 1]
```diff
@@ -89,7 +89,7 @@
     public final static String INVALID_INSTANCE = "invalid_instance";
 
     /**
-     * Request to server failed, but no error and error_description is returned back from the service。
+     * Request to server failed, but no error and error_description is returned back from the service.
```
Comment (jaredsburrows) [inline msal/src/main/java/com/microsoft/identity/client/MsalServiceException.java]: Bad windows character broke the build for linux/mac locally.

**Comment by jaredsburrows** [comment]
@iambmelt No problem. Thank you!

File: msal/src/cache/java/com/microsoft/identity/client/BaseTokenCacheItem.java [hunk 2]
```diff
@@ -31,13 +31,13 @@
 abstract class BaseTokenCacheItem {
 
     @SerializedName("client_id")
-    final String mClientId;
+    String mClientId;
```
Comment (weijjia) [inline msal/src/cache/java/com/microsoft/identity/client/BaseTokenCacheItem.java]: i'm kind of against removing final. Final guarantees that those variables are only initialized in the constructor.

Comment (jaredsburrows) [inline msal/src/cache/java/com/microsoft/identity/client/BaseTokenCacheItem.java]: @weijjia You have to for the empty constructor. Plus, those values are only initialized if the server returns those values else they will be null.

Comment (weijjia) [inline msal/src/cache/java/com/microsoft/identity/client/BaseTokenCacheItem.java]: why you cannot do this(null, null, null) which calls the other constructor?

Comment (jaredsburrows) [inline msal/src/cache/java/com/microsoft/identity/client/BaseTokenCacheItem.java]: @weijjia If they are not initialized and no data comes back(no json), then it will be `null` anyways. You want me to add more code to just "make" them `null`?

@iambmelt @nazukj Based on the documentation example, they are not using final: https://static.javadoc.io/com.google.code.gson/gson/2.8.1/com/google/gson/annotations/SerializedName.html.

Also, if you finalize the parameters, I believe GSON is forced to use reflection like this:  https://stackoverflow.com/a/11220912/950427. 

I am trying to keep this pull request similar to the one I made for ADAL: https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/842

Comment (weijjia) [inline msal/src/cache/java/com/microsoft/identity/client/BaseTokenCacheItem.java]: I'll find out how GSON will do it internally and let you know.

Comment (jaredsburrows) [inline msal/src/cache/java/com/microsoft/identity/client/BaseTokenCacheItem.java]: @iambmelt @weijjia This PR solves the unsafe allocator issue: https://github.com/google/gson/blob/09cbc8073bb2bc60aeed6b831be4473daaf915ae/gson/src/main/java/com/google/gson/internal/UnsafeAllocator.java#L42.

To prevent introducing a new one, I removed the finals: https://github.com/google/gson/blob/9e44d60b83d667d96150eeff4e3f04cd4f157e9a/gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java#L157.

The `ReflectiveTypeAdapterFactory` is used in the `Gson` class here: https://github.com/google/gson/blob/92b52d25d844b81620a04917e5fa164d44b9c8a5/gson/src/main/java/com/google/gson/Gson.java#L252.

Basically I would like this to use as little of reflection as we can.

Comment (weijjia) [inline msal/src/cache/java/com/microsoft/identity/client/BaseTokenCacheItem.java]: I'll need time to investigate into it. We should finalize everything if we can and of course reduce reflections. I'll get back to the PR on Monday.

**Comment by jaredsburrows** [comment]
@weijjia 
@nazukj @addev-ashish Any update on moving forward with my optimizations?

**Comment by weijjia** [comment]
If I'm reading the code correctly, unSafeAllocator will be used if no default constructor is provided. So your code change about providing will solve that issue(Thanks so much for sending the PR). https://github.com/google/gson/blob/0636635cbffa08157bdbd558b1212e4d806474eb/gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java#L94. 

Then we come down why we have to remove the final. I'm not sure i fully understand this part, 
https://github.com/google/gson/blob/0636635cbffa08157bdbd558b1212e4d806474eb/gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java#L46
this is how the type adapter is chosen, how does it help remove the reflection?

**Comment by jaredsburrows** [comment]
@weijjia Not entirely remove it but it forces Gson to check if the "member is accessible" and since the field is `final` it has to `setAccessible(true)` here: https://github.com/google/gson/blob/9e44d60b83d667d96150eeff4e3f04cd4f157e9a/gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java#L157.

**Comment by weijjia** [comment]
Sorry if i understood, but this is setting it anyways. getBoundFields is used in the reflective adapter, whenever this adapter is used, reflection is not avoid at all.

**Comment by jaredsburrows** [comment]
@weijjia Honestly, this is taking much longer than it should be. I just wanted to make sure this library was using the "no-args" constructor.

I can add the `final`s back in, I just believe that the Gson API has to use reflection to make them accessible. I just want to use reflection as little as I can.

**Comment by jaredsburrows** [comment]
@weijjia I have added in the `final` for the fields. The PR should be ready to go now!
@nazukj @addev-ashish

**Comment by jaredsburrows** [comment]
@weijjia Adding `this(null)` to the constructors like you asked seems to break your build:
```
com.microsoft.identity.client.InteractiveRequestTest > testAcquireTokenExpiresInNotReturned[test(AVD) - 5.0.2] FAILED 
	java.lang.RuntimeException: Failed to invoke com.microsoft.identity.client.AccessTokenCacheItem() throws com.microsoft.identity.client.MsalClientException with no args
	at com.google.gson.internal.ConstructorConstructor$3.construct(ConstructorConstructor.java:116)
Tests on test(AVD) - 5.0.2 failed: Instrumentation run failed due to 'java.lang.NullPointerException'
:msal:connectedDebugAndroidTest FAILED
```

**Comment by jaredsburrows** [comment]
@weijjia Adding back in my original changes. Everything passes! Thank you.

**Comment by jaredsburrows** [comment]
@weijjia Thank you for your support!
